### DISCUSSION TITLE
Model becomes worse at detecting instances after additional training rounds

### DISCUSSION THREAD

**Discussion by adenar95** [talmolab/sleap#1340] [Help!]
Hi everyone,

I have been working with SLEAP for a few weeks and really like the software a lot!
I am working on videos of flies and it works quite well, although I have noticed, that after the 3rd round of training the inference on new images got worse when I compared it with the model after the 2nd round of training. I have 8 different cameras and some different treatments that slightly alter the background, so I tried to represent all these things in the training data. My training was done as follows: There are 16 flies in each to be detected.
1st round: 20 frames per camera (20x8), each treatment represented in a total of 40 frames. Total number of frames 160.
2nd round: additional 40 frames per camera (40x8), Total number of frames 160 (from first labeling round) + 320 = 480
3rd round: additional 60 frames per camera (60x8), total number of frames 160 + 320 + 480= 960

I am using the multi-animal top-down approach.

What could be the reason for the model becoming worse? I checked the curves for training loss and validation loss (attached) and it doesn't look like overfitting. Also the inference worked equally bad with tracker set to simple, flow or none. 
![Screenshot 2023-06-06 at 19 33 12](https://github.com/talmolab/sleap/assets/135736891/491bf42c-665c-4ded-b938-2405345629c4)

I would really really appreciate some inputs from the community.. Thanks so much in advance!

Best,

adenar

**Comment by roomrys** [comment]
Hi @adenar95,

I am wondering if you are using a fixed set of frames for the validation set? This would be required to perform an accurate model analysis.

Unless you are specifying the same validation set across training rounds (only possible by setting the `validation_labels` in the `training_config.json` and then running training through the `sleap-train` CLI), SLEAP will use a fraction of your annotated frames as the validation set. This validation set changes as more labels are added. Moreover, if you start labeling more difficult frames as you progress along the project, you will notice that SLEAP will also have more difficulty predicting on these difficult frames.

Let us know if this sounds like it might be the case, or if you think something else is going on.

Thanks,
Liezl

**Reply by adenar95** [reply]
Hi roomrys,

thanks for your reply. No actually my `validation_labels ` setting is not set to a specific set. What sort of frames would you choose as a validation set? same representation of different camers etc. as in training set, but unseen pictures? and how many?

The flies that are not recognized in the inference after the third round of training are actually instances that were easily detected in previous rounds and not in difficult poses or hardly visible.. so I don't know if either of these could be the problem?

Thanks!
Alessio

**Reply by roomrys** [reply]
Hi @adenar95,

Ah ok, my answer above is for quantitative model analysis (comparing metrics across models) which would likely be best done when you have a final model you are happy with - or more importantly, all the labeled frames that were used to train this final model (see *Quantitative Model Analysis* if you are interested). 

I would check the model hyperparameters (has the max stride, input scaling, min/max rotation changed across models). The model hyperparameters can greatly affect results. Most notable would be the receptive field size (this should contain the entire animal). And, of course, the newer predictions will be affected by the new labels. You can check that the quality of your labels has not decreased as more labels have been added, but I would check the model parameters first.

Let me know if this helps!

Thanks,
Liezl

---

### Quantitative Model Analysis
This is a bit complex, but is just an option for down the road if you choose...

After you have finished annotating your project, you can use the `sleap.nn.data.training.split_labels_train_val` function to split your final `Labels` into two sets of `Labels`: the training `Labels` and the validation `Labels`.
https://github.com/talmolab/sleap/blob/1a0404c0ffae7b248eb360562b0bb95a42a287b6/sleap/nn/data/training.py#L14-L41
> You can further split the training set into smaller training sets to measure how increasing the number of labeled frames affects model performance. Similarly, you can split your validation set into a validation and test set where the test set is never seen by the model during training and could be used to measure unbiased model performance.

You would then need to save both sets (or all sets- if you further split the `Labels`) to an slp and specify the paths to the slp in your `training_config.json`. 
> You may want to save a copy of the `training_config.json` before making changes - for reproducibility purposes!

Namely, you would locate the `training_config.json` for the model you would like to re-train (and gather metrics for) and specify the `training_labels` and `validation_labels` (and optionally the `test_labels`). I've highlighted the section of the `training_config` from a default profile as an example:
https://github.com/talmolab/sleap/blob/1a0404c0ffae7b248eb360562b0bb95a42a287b6/tests/data/models/min_tracks_2node.UNet.topdown_multiclass/training_config.json#L3-L14
> You may notice that there are a few different fields that can specify which labeled frames to use for training/validation/test. Each time you train a model, the `training_job.config` will save a list on indices in `training_inds` to keep track of precisely which frames in the project were used to train the model. Thus, if you wanted to use the same labeled frames for your model analysis, you would instead populate the {`training`, `validation`, `test`}`_inds` sections with the list of indices output from `split_labels_train_val`. You would then ommit the {`training`, `validation`, `test`}`_labels` sections and set `split_by_inds` to True. This might get a bit hairy as you would need to exclude the indices used for training (across all models)....

Finally! The last step is to form and run your [`sleap-train`](https://sleap.ai/guides/cli.html#sleap-train) command via the terminal.

**Reply by adenar95** [reply]
Hi roomrys,

thanks again for all your valuable answers!

So I ran the three rounds of training again with the above specified number of frames for each round and with an unchanging independent validation set of a 160 frames, also from all the different cameras. 

I think I also learned a couple of other things from your answers that I payed more attention to. Now I find that the training logs for the centroid model (loss and validation loss plotted against epochs) it looks like this
![Screenshot 2023-06-14 at 15 43 23](https://github.com/talmolab/sleap/assets/135736891/e857571d-57fa-443f-8c5e-8d898cf98d41)

and here the same for the centered instance model:
![Screenshot 2023-06-14 at 15 40 02](https://github.com/talmolab/sleap/assets/135736891/001ff7a1-e86c-425f-8ab5-aac6995eb2f3)

I see that the validation loss becomes slightly better, but the training loss becomes worse rather than better. What could be the problem? And can I expect to get further improvements with additional rounds of training or does it look like it's not worth adding more training data?

Thanks for some more advice..

best,

adenar

P.S.: I was just wondering after posting this, if in the second round of training I should only train the model on the new frames but remove the ones from the first round from the training data? or should I do it as I did, with additional frames but leave the old ones in? same for the third round.. would the model learn better on only new frames?

**Comment by roomrys** [comment]
Hi @adenar95,

I would interpret the curves to mean that more diverse frames are leading to general difficulty predicting. It is good that the validation is decreasing as we are doing better on images the network hasn't directly seen (although the validation loss is still used to tune the network and the choice of validation set can subsequently affect training). It is possible to overfit, so if you are resuming training all the time, I would give it a shot with training from scratch (uncheck resume training).

If you choose to resume training of your model, but exclude frames, then the model will be more heavily tuned towards the examples you included. No right or wrong, just different based on how you want to do your training. However, if you are not resuming training, then the new model will only have ever seen the "new" examples and thus be a new model.

If you are unhappy with the results, then usually it can be corrected by adding more examples (or going back and reviewing the examples you've already annotated). I personally use the qualitative method you mentioned earlier to determine when I should stop labeling. But if adding/correcting examples doesn't seem to help, I would play with the model hyperparameters - as mentioned above. Or, if you are over labeling, then there are some things that can be smoothed out during post-processing (such as jumping nodes).

Thanks,
Liezl
